src/libraries/System.Private.CoreLib/docs/Container Design Notes.md

```diff
@@ -0,0 +1,60 @@
+Clear operations are also legal during enumeration in addition to Remove. TrimExcess, however, is *not* legal, and will invalidate enumeration.
+
+### Non-randomized String Comparers Transition to Randomized After Collisions
+
+By default, Dictionary will substitute high-performance 'non-randomized string comparers' for the built-in string comparers if provided one of them at construction time. After precisely 100 hash collisions (see `System.Collections.HashHelpers.HashCollisionThreshold`) the container will switch to the originally-provided randomized comparer and rehash, re-generating its hashcodes and buckets. This does not affect the contents of the entries table so it will not violate the consistent ordering guarantee listed above.
+
+This means that the comparer actually being used by the container is not necessarily `Dictionary.Comparer` but is instead a hidden internal comparer, `Dictionary._comparer`. It is important to preserve the rule that `Dictionary.get_Comparer` returns the comparer that was provided when the container was constructed, and ensure that the use of non-randomized comparers is not observable from outside the container.
+
+### HashCode Caching
+
+Dictionary caches the result of `GetHashCode` for a key at insertion time inside of the `Entry` representing the key-value pair, which means that rehashing operations (with the exception of the rehash triggered by a transition to the randomized string comparers mentioned above) will not invoke `GetHashCode`. Dictionary also makes sure to compare the cached hash against the pre-computed hash of the target key during lookups, removals and insertions before calling `Equals` on the provided comparer. These behaviors are observable by end users and covered by tests, but are not guaranteed by documentation.
+
+Passing a Dictionary of identical type to the Dictionary constructor will reuse the existing `_entries` array from the dictionary being copied if possible, so this operation also does not invoke `GetHashCode` when the source and destination comparer are identical.
+
+Documentation states that any key used in a Dictionary must not change in any way that affects its hash after insertion, so it is technically possible to remove or change this caching behavior in the future, but if a user provides a comparer with side-effects, such a change could cause breakage.
+
+### Thread-safety Violations Must Produce Exceptions
+
+Dictionary is designed to reliably produce exceptions when it is manipulated from multiple threads at once without synchronization, and we have multiple tests that verify this. "Improving" Dictionary to be able to perform operations safely across multiple threads without synchronization or throwing will as a result produce test failures and could result in observable behavior changes.
```

Comment:
These tests are really only testing that the thread safety violations do not lead to hangs.